File: src/sarma_cli/session.py
"""Session lifecycle — wraps IDE agent runtime for CLI use.

Workflow-aware: reads the current workflow from the registry on each turn
to determine execution mode (ruflo vs audit pipeline).
"""

from __future__ import annotations

import uuid
from typing import Any, AsyncIterator

from sarma_cli.engine.agent_factory import AgentFactory
from sarma_cli.engine.agent_runner import AgentRunner
from sarma_cli.engine.mcp_pool import McpClientPool
from sarma_cli.engine.models import ConversationMessage, StreamEvent
from sarma_cli.engine.enums import StreamEventType
from sarma_cli.context.compaction import (
    STRUCTURED_MEMORY_PROMPT,
    ContextCompactor,
    ContextWindowPolicy,
    estimate_static_prompt_tokens,
)

from sarma_cli.config import CliConfig
from sarma_cli.runtime.resolver import RuntimePolicyResolver
from sarma_cli.store import Store


class Session:
    """Manages a conversation with the LangGraph agent pipeline.

    Workflow-aware: reads the current workflow from the registry on each
    turn to determine execution mode. Switching /workflow takes effect on
    the next turn — no session restart needed.
    """

    def __init__(self, config: CliConfig, store: Store) -> None:
        self._config = config
        self._store = store
        self._pool = McpClientPool()
        self._factory = AgentFactory(self._pool)
        self._resolver = RuntimePolicyResolver(config)
        self._conversation_id: str = ""
        self._history: list[ConversationMessage] = []
        self._graph_state: dict[str, Any] = {
            "current_stage": "",
            "completed": set(),
            "failed": None,
            "gapfill_loops": 0,
            "feedback_loops": 0,
        }
        self._compact_trigger_ratio = 0.90
        self._compact_target_ratio = 0.55

    @property
    def graph_state(self) -> dict[str, Any]:
        return dict(self._graph_state)

    @property
    def conversation_id(self) -> str:
        return self._conversation_id

    @property
    def pool(self) -> McpClientPool:
        return self._pool

    @property
    def tool_count(self) -> int:
        return len(self._pool.tools)

    async def ensure_mcp_connected(self, workflow: str) -> None:
        """Connect the runtime MCP pool for the given workflow without building an agent."""
        run_plan = self._resolver.resolve(workflow)
        server_configs = {
            server.name: server.to_langchain_config()
            for server in run_plan.enabled_servers
        }
        await self._pool.connect(server_configs)

    async def restart_runtime(self) -> None:
        """Rebuild runtime resources while preserving conversation history."""
        await self._pool.disconnect()
        self._pool = McpClientPool()
        self._factory = AgentFactory(self._pool)
        self._resolver = RuntimePolicyResolver(self._config)
        self._reset_graph_state()

    def new_conversation(self, title: str = "") -> str:
        from sarma_cli.workflows import get_registry
        mode = get_registry().current_name()
        provider = self._resolver.provider_for(mode or "ruflo")
        self._conversation_id = self._store.create_conversation(
            title=title or f"{mode.title()} session",
            model_name=provider.model_name,
        )
        self._history.clear()
        self._reset_graph_state()
        return self._conversation_id

    def resume_conversation(self, cid: str) -> bool:
        messages = self._store.load_messages(cid)
        if not messages:
            return False
        self._conversation_id = cid
        self._history = [
            ConversationMessage(
                id=m["id"],
                conversation_id=cid,
                turn_id=m["turn_id"],
                role=m["role"],
                content=m["content"],
                reasoning_content=m.get("reasoning"),
            )
            for m in messages
        ]
        return True

    async def compact_context(
        self,
        *,
        force: bool = True,
        workflow: str = "ruflo",
        upcoming_text: str = "",
        system_prompt: str = "",
    ) -> bool:
        """Compact history into structured memory near the context limit."""
        compactor = self._compactor_for_workflow(
            workflow,
            system_prompt=system_prompt,
        )
        changed, new_history, memory = await compactor.compact(
            self._history,
            lambda messages: self._summarize_messages(messages, workflow=workflow),
            conversation_id=self._conversation_id,
            upcoming_text=upcoming_text,
            force=force,
        )
        if not changed:
            return False

        source_count = len(self._history) - (len(new_history) - 1)
        self._history = new_history
        if self._conversation_id:
            self._store.save_memory_artifact(
                self._conversation_id,
                memory,
                source_count=max(source_count, 0),
            )
            self._store.replace_messages(self._conversation_id, self._history)
        return True

    async def run_turn(self, user_message: str) -> AsyncIterator[StreamEvent]:
        """Execute one turn, yielding StreamEvents.

        Reads the current workflow from the registry to determine execution
        mode. Switching /workflow changes what runs on the next turn.
        """
        from sarma_cli.workflows import get_registry

        # Determine mode from current workflow
        registry = get_registry()
        current_wf = registry.current()
        mode = current_wf.name if current_wf else "ruflo"
        if not self._conversation_id:
            self.new_conversation(title=user_message[:60])

        turn_id = uuid.uuid4().hex[:12]
        run_plan = self._resolver.resolve(mode)

        await self.compact_context(
            force=False,
            workflow=mode,
            upcoming_text=user_message,
            system_prompt=run_plan.system_prompt,
        )

        runner_history = list(self._history)

        # Persist user message
        self._store.save_message(
            self._conversation_id, turn_id, "user", user_message
        )
        self._history.append(ConversationMessage(
            role="user", content=user_message,
            conversation_id=self._conversation_id, turn_id=turn_id,
        ))

        runner = AgentRunner(
            factory=self._factory,
            pool=self._pool,
            provider=run_plan.provider,
            enabled_servers=run_plan.enabled_servers,
            skill=run_plan.skill,
            history=runner_history,
            system_prompt=run_plan.system_prompt,
            conversation_id=self._conversation_id,
            turn_id=turn_id,
            mode=mode,
            subagent_providers=run_plan.subagent_providers,
            subagent_mcp_allow=run_plan.subagent_mcp_allow,
            subagent_skills=run_plan.subagent_skills,
        )

        async for event in runner.run(user_message):
            if mode in ("audit", "audit-slim"):
                self._track_graph_progress(event)
            yield event

        # Persist assistant response
        if runner.assistant_content:
            self._store.save_message(
                self._conversation_id, turn_id, "assistant",
                runner.assistant_content,
                reasoning=runner.reasoning_content or None,
            )
            self._history.append(ConversationMessage(
                role="assistant", content=runner.assistant_content,
                conversation_id=self._conversation_id, turn_id=turn_id,
                reasoning_content=runner.reasoning_content or None,
            ))
            update_fields: dict[str, Any] = {"status": "idle"}
            # Set a title from the first user message only if none was set yet.
            existing = self._store.get_conversation(self._conversation_id)
            if existing is not None and not (existing.get("title") or "").strip():
                update_fields["title"] = user_message[:60] or "Untitled session"
            self._store.update_conversation(self._conversation_id, **update_fields)

    def _reset_graph_state(self) -> None:
        self._graph_state = {
            "current_stage": "",
            "completed": set(),
            "failed": None,
            "gapfill_loops": 0,
            "feedback_loops": 0,
        }

    def _should_compact(self, budget: int, *, upcoming_text: str = "") -> bool:
        compactor = ContextCompactor(ContextWindowPolicy(
            max_context_tokens=budget,
            trigger_ratio=self._compact_trigger_ratio,
            raw_tail_ratio=self._compact_target_ratio,
        ))
        return compactor.plan(
            self._history,
            upcoming_text=upcoming_text,
        ).should_compact

    def _context_budget(self, workflow: str) -> int:
        provider = self._resolver.provider_for(workflow)
        return max(int(provider.max_context_tokens or 128_000), 1)

    def _split_for_compaction(
        self,
        budget: int,
    ) -> tuple[list[ConversationMessage], list[ConversationMessage]]:
        return ContextCompactor(ContextWindowPolicy(
            max_context_tokens=budget,
            trigger_ratio=self._compact_trigger_ratio,
            raw_tail_ratio=self._compact_target_ratio,
        )).split_raw_tail(self._history)

    def _estimate_history_tokens(self, messages: list[ConversationMessage]) -> int:
        return ContextCompactor(ContextWindowPolicy(
            max_context_tokens=128_000,
        )).estimate_history_tokens(messages)

    @staticmethod
    def _estimate_message_tokens(message: ConversationMessage) -> int:
        return ContextCompactor.estimate_message_tokens(message)

    @staticmethod
    def _estimate_text_tokens(text: str) -> int:
        return ContextCompactor.estimate_text_tokens(text)

    def _compactor_for_workflow(
        self,
        workflow: str,
        *,
        system_prompt: str = "",
    ) -> ContextCompactor:
        return ContextCompactor(ContextWindowPolicy(
            max_context_tokens=self._context_budget(workflow),
            trigger_ratio=self._compact_trigger_ratio,
            raw_tail_ratio=self._compact_target_ratio,
            static_prompt_tokens=estimate_static_prompt_tokens(
                system_prompt,
                self.tool_count,
            ),
        ))

    async def _summarize_messages(
        self,
        messages: list[ConversationMessage],
        *,
        workflow: str = "ruflo",
    ) -> str:
        from langchain_core.messages import HumanMessage, SystemMessage

        provider = self._resolver.provider_for(workflow)
        model = self._factory._init_model(provider, None)
        transcript = "\n\n".join(
            f"{message.role.upper()}: {message.content}"
            for message in messages
            if message.content
        )
        result = await model.ainvoke([
            SystemMessage(content=STRUCTURED_MEMORY_PROMPT),
            HumanMessage(content=transcript),
        ])
        return str(getattr(result, "content", "") or "")

    def _track_graph_progress(self, event: StreamEvent) -> None:
        """Update graph state from subagent lifecycle events (audit mode only)."""
        etype = event.type
        payload = event.payload

        if etype == StreamEventType.SUBAGENT_START:
            name = payload.get("subagent", "")
            if name:
                self._graph_state["current_stage"] = name

        elif etype == StreamEventType.SUBAGENT_COMPLETE:
            name = payload.get("subagent", "")
            if name:
                self._graph_state["completed"].add(name)
                if name == self._graph_state.get("current_stage"):
                    self._graph_state["current_stage"] = ""

        elif etype == StreamEventType.SUBAGENT_ERROR:
            name = payload.get("subagent", "")
            if name:
                self._graph_state["failed"] = name

        elif etype == StreamEventType.CUSTOM_PROGRESS:
            data = payload.get("data", {})
            if isinstance(data, dict) and data.get("type") == "audit_route":
                loop = data.get("loop", "")
                count = data.get("count")
                if loop == "gapfill" and count is not None:
                    self._graph_state["gapfill_loops"] = int(count)
                elif loop == "feedback" and count is not None:
                    self._graph_state["feedback_loops"] = int(count)

    async def close(self) -> None:
        await self._pool.disconnect()

File: src/session.ts
/**
 * Session lifecycle — wraps the agent runtime for CLI use.
 *
 * Workflow-aware: reads the current workflow on each turn to determine
 * execution mode (ruflo vs audit pipeline). Switching the workflow takes
 * effect on the next turn — no session restart needed.
 */

import { randomUUID } from "node:crypto";
import { SystemMessage, HumanMessage } from "@langchain/core/messages";

import { AgentFactory } from "@/engine/agentFactory";
import { AgentRunner } from "@/engine/agentRunner";
import { McpClientPool } from "@/engine/mcpPool";
import { ConversationMessage, StreamEvent } from "@/engine/models";
import { ModelFactory } from "@/engine/modelFactory";
import { ModelProviderDTO } from "@/engine/dto";
import { StreamEventType } from "@/engine/enums";
import {
  makeRunCompletedEvent,
  makeRunFailedEvent,
  makeRunStartedEvent,
} from "@/engine/streaming";
import {
  STRUCTURED_MEMORY_PROMPT,
  ContextCompactor,
  ContextWindowPolicy,
  estimateStaticPromptTokens,
} from "@/context/compaction";
import { createTokenEstimator } from "@/context/tokenizer";
import { type CliConfig, type ProviderConfig } from "@/config";
import { RuntimePolicyResolver } from "@/runtime/resolver";
import { AgentRuntimeServices } from "@/runtime/services";
import { Store } from "@/store";
import { defaultWorkflowName, getWorkflowMeta } from "@/workflows";

export interface GraphState {
  current_stage: string;
  completed: Set<string>;
  failed: string | null;
  gapfill_loops: number;
  feedback_loops: number;
}

function uid(): string {
  return randomUUID().replace(/-/g, "").slice(0, 12);
}

/** Manages a conversation with the LangGraph agent pipeline. */
export class Session {
  private pool: McpClientPool;
  private modelFactory: ModelFactory;
  private runtimeServices: AgentRuntimeServices;
  private factory: AgentFactory;
  private resolver: RuntimePolicyResolver;
  private _conversationId = "";
  private _history: ConversationMessage[] = [];
  private _graphState: GraphState = Session.freshGraphState();
  private readonly compactTriggerRatio = 0.9;
  private readonly compactTargetRatio = 0.55;
  private currentWorkflow: string = defaultWorkflowName();
  private currentRunAbort: AbortController | null = null;

  constructor(
    private readonly config: CliConfig,
    private readonly store: Store,
  ) {
    this.pool = new McpClientPool();
    this.modelFactory = new ModelFactory();
    this.runtimeServices = AgentRuntimeServices.create();
    this.factory = new AgentFactory(this.pool, {
      modelFactory: this.modelFactory,
      runtimeServices: this.runtimeServices,
    });
    this.resolver = new RuntimePolicyResolver(config);
  }

  private static freshGraphState(): GraphState {
    return {
      current_stage: "",
      completed: new Set(),
      failed: null,
      gapfill_loops: 0,
      feedback_loops: 0,
    };
  }

  get graphState(): GraphState {
    return {
      ...this._graphState,
      completed: new Set(this._graphState.completed),
    };
  }

  get conversationId(): string {
    return this._conversationId;
  }

  get poolRef(): McpClientPool {
    return this.pool;
  }

  get toolCount(): number {
    return this.pool.tools.length;
  }

  /** Set the active workflow that the next turn will run. */
  setWorkflow(name: string): void {
    if (!getWorkflowMeta(name)) {
      throw new Error(`Unknown workflow: ${name}`);
    }
    this.currentWorkflow = name;
  }

  get workflow(): string {
    return this.currentWorkflow;
  }

  /** Connect the runtime MCP pool for the given workflow without building an agent. */
  async ensureMcpConnected(workflow: string): Promise<void> {
    const runPlan = this.resolver.resolve(workflow);
    await this.pool.connect(Session.serverConfigsFor(runPlan.enabledServers));
  }

  /** Rebuild runtime resources while preserving conversation history. */
  async restartRuntime(): Promise<void> {
    this.cancelCurrentRun();
    await this.pool.disconnect();
    this.pool = new McpClientPool();
    this.modelFactory = new ModelFactory();
    this.runtimeServices = AgentRuntimeServices.create();
    this.factory = new AgentFactory(this.pool, {
      modelFactory: this.modelFactory,
      runtimeServices: this.runtimeServices,
    });
    this.resolver = new RuntimePolicyResolver(this.config);
    this.resetGraphState();
  }

  cancelCurrentRun(): boolean {
    const controller = this.currentRunAbort;
    if (!controller || controller.signal.aborted) return false;
    controller.abort();
    return true;
  }

  newConversation(title = ""): string {
    const mode = this.currentWorkflow;
    const provider = this.resolver.providerFor(mode || "ruflo");
    const titledMode = mode.charAt(0).toUpperCase() + mode.slice(1);
    this._conversationId = this.store.createConversation(
      title || `${titledMode} session`,
      provider.modelName,
    );
    this._history = [];
    this.resetGraphState();
    return this._conversationId;
  }

  resumeConversation(cid: string): boolean {
    const messages = this.store.loadMessages(cid);
    if (messages.length === 0) return false;
    this._conversationId = cid;
    this._history = messages.map(
      (m) =>
        new ConversationMessage({
          id: m.id,
          conversationId: cid,
          turnId: m.turn_id,
          role: m.role,
          content: m.content,
          reasoningContent: m.reasoning ?? null,
        }),
    );
    return true;
  }

  /** Compact history into structured memory near the context limit. */
  async compactContext(
    options: { force?: boolean; workflow?: string; upcomingText?: string; systemPrompt?: string } = {},
  ): Promise<boolean> {
    const workflow = options.workflow ?? "ruflo";
    const compactor = this.compactorForWorkflow(workflow, options.systemPrompt ?? "");
    const [changed, newHistory, memory] = await compactor.compact(
      this._history,
      (messages) => this.summarizeMessages(messages, workflow),
      {
        conversationId: this._conversationId,
        upcomingText: options.upcomingText ?? "",
        force: options.force ?? true,
      },
    );
    if (!changed) return false;

    const sourceCount = this._history.length - (newHistory.length - 1);
    this._history = newHistory;
    if (this._conversationId) {
      this.store.saveMemoryArtifact(this._conversationId, memory, "structured_context", Math.max(sourceCount, 0));
      this.store.replaceMessages(
        this._conversationId,
        this._history.map((m) => ({
          id: m.id ?? undefined,
          turn_id: m.turnId,
          role: m.role,
          content: m.content,
          reasoning_content: m.reasoningContent ?? null,
          created_at: m.createdAt ?? undefined,
        })),
      );
    }
    return true;
  }

  /** Execute one turn, yielding StreamEvents. */
  async *runTurn(userMessage: string): AsyncIterableIterator<StreamEvent> {
    const mode = this.currentWorkflow || "ruflo";
    if (!getWorkflowMeta(mode)) {
      throw new Error(`Unknown workflow: ${mode}`);
    }
    if (!this._conversationId) {
      this.newConversation(userMessage.slice(0, 60));
    }

    const turnId = uid();
    const runPlan = this.resolver.resolve(mode);
    const toolExecutionIds = new Map<string, string>();
    const abortController = new AbortController();
    this.currentRunAbort = abortController;

    yield makeRunStartedEvent(this._conversationId, turnId);

    try {
      if (abortController.signal.aborted) throw new Error("Run cancelled.");
      await this.pool.connect(Session.serverConfigsFor(runPlan.enabledServers));
      if (abortController.signal.aborted) throw new Error("Run cancelled.");

      await this.compactContext({
        force: false,
        workflow: mode,
        upcomingText: userMessage,
        systemPrompt: runPlan.systemPrompt,
      });

      const runnerHistory = [...this._history];

      // Persist user message.
      this.store.saveMessage(this._conversationId, turnId, "user", userMessage);
      this._history.push(
        new ConversationMessage({
          role: "user",
          content: userMessage,
          conversationId: this._conversationId,
          turnId,
        }),
      );

      const runner = new AgentRunner({
        factory: this.factory,
        pool: this.pool,
        provider: runPlan.provider,
        enabledServers: runPlan.enabledServers,
        skill: runPlan.skill,
        history: runnerHistory,
        systemPrompt: runPlan.systemPrompt,
        conversationId: this._conversationId,
        turnId,
        mode,
        subagentProviders: runPlan.subagentProviders,
        subagentMcpAllow: runPlan.subagentMcpAllow,
        subagentSkills: runPlan.subagentSkills,
        rag: runPlan.rag,
        abortSignal: abortController.signal,
      });

      for await (const event of runner.run(userMessage)) {
        if (abortController.signal.aborted) throw new Error("Run cancelled.");
        this.persistToolEvent(event, toolExecutionIds);
        if (mode === "audit" || mode === "audit-slim") {
          this.trackGraphProgress(event);
        }
        yield event;
      }

      // Persist assistant response. For audit modes finalContent is the report
      // stage; for Ruflo it is the accumulated assistant tokens.
      const finalContent = runner.finalContent;
      if (finalContent) {
        this.store.saveMessage(
          this._conversationId,
          turnId,
          "assistant",
          finalContent,
          null,
          runner.reasoningContent || null,
        );
        this._history.push(
          new ConversationMessage({
            role: "assistant",
            content: finalContent,
            conversationId: this._conversationId,
            turnId,
            reasoningContent: runner.reasoningContent || null,
          }),
        );
        const updateFields: Record<string, string> = { status: "idle" };
        const existing = this.store.getConversation(this._conversationId);
        if (existing && !(existing.title || "").trim()) {
          updateFields.title = userMessage.slice(0, 60) || "Untitled session";
        }
        this.store.updateConversation(this._conversationId, updateFields);
      }
      yield makeRunCompletedEvent(this._conversationId, turnId, finalContent);
    } catch (exc) {
      const message = abortController.signal.aborted ? "Run cancelled." : exc instanceof Error ? exc.message : String(exc);
      yield makeRunFailedEvent(this._conversationId, turnId, message);
    } finally {
      if (this.currentRunAbort === abortController) this.currentRunAbort = null;
    }
  }

  private resetGraphState(): void {
    this._graphState = Session.freshGraphState();
  }

  private contextBudget(workflow: string): number {
    const provider = this.resolver.providerFor(workflow);
    return Math.max(Math.trunc(provider.maxContextTokens || 128_000), 1);
  }

  private compactorForWorkflow(workflow: string, systemPrompt = ""): ContextCompactor {
    const provider = this.resolver.providerFor(workflow);
    const estimateText = createTokenEstimator(provider);
    return new ContextCompactor(
      new ContextWindowPolicy({
        maxContextTokens: Math.max(Math.trunc(provider.maxContextTokens || 128_000), 1),
        triggerRatio: this.compactTriggerRatio,
        rawTailRatio: this.compactTargetRatio,
        staticPromptTokens: estimateStaticPromptTokens(
          systemPrompt,
          this.toolCount + Session.builtinAndMiddlewareToolCount(this.resolver.resolve(workflow).rag),
          estimateText,
        ),
        tokenEstimator: estimateText,
      }),
    );
  }

  private async summarizeMessages(
    messages: ConversationMessage[],
    workflow = "ruflo",
  ): Promise<string> {
    const provider: ProviderConfig = this.resolver.providerFor(workflow);
    const model = this.modelFactory.initModel(providerToDto(provider), null);
    const transcript = messages
      .filter((m) => m.content)
      .map((m) => `${m.role.toUpperCase()}: ${m.content}`)
      .join("\n\n");
    const result = await model.invoke([
      new SystemMessage({ content: STRUCTURED_MEMORY_PROMPT }),
      new HumanMessage({ content: transcript }),
    ]);
    const content = (result as { content?: unknown }).content;
    return typeof content === "string" ? content : String(content ?? "");
  }

  private trackGraphProgress(event: StreamEvent): void {
    const etype = event.type;
    const payload = event.payload;

    if (etype === StreamEventType.STAGE_START) {
      const name = (payload.stage as string) || (payload.subagent as string) || "";
      if (name) this._graphState.current_stage = name;
    } else if (etype === StreamEventType.STAGE_COMPLETE) {
      const name = (payload.stage as string) || (payload.subagent as string) || "";
      if (name) {
        this._graphState.completed.add(name);
        if (name === this._graphState.current_stage) this._graphState.current_stage = "";
      }
    } else if (etype === StreamEventType.STAGE_ERROR) {
      const name = (payload.stage as string) || (payload.subagent as string) || "";
      if (name) this._graphState.failed = name;
    } else if (etype === StreamEventType.CUSTOM_PROGRESS) {
      const data = payload.data;
      if (data && typeof data === "object" && (data as { type?: string }).type === "audit_route") {
        const loop = (data as { loop?: string }).loop ?? "";
        const count = (data as { count?: number }).count;
        if (loop === "gapfill" && count !== undefined && count !== null) {
          this._graphState.gapfill_loops = Math.trunc(count);
        } else if (loop === "feedback" && count !== undefined && count !== null) {
          this._graphState.feedback_loops = Math.trunc(count);
        }
      }
    }
  }

  private persistToolEvent(event: StreamEvent, toolExecutionIds: Map<string, string>): void {
    if (!this._conversationId) return;
    const payload = event.payload;
    if (event.type === StreamEventType.TOOL_START) {
      const toolName = String(payload.tool_name ?? "");
      const callId = String(payload.tool_call_id ?? "");
      const argsJson = String(payload.args_json ?? JSON.stringify(payload.args ?? {}));
      const serverName = toolName.includes("__") ? toolName.split("__", 1)[0]! : "";
      const id = this.store.saveToolExecution(this._conversationId, event.turnId, toolName, argsJson, serverName);
      if (callId) toolExecutionIds.set(callId, id);
      return;
    }
    if (event.type === StreamEventType.TOOL_RESULT || event.type === StreamEventType.TOOL_ERROR) {
      const callId = String(payload.tool_call_id ?? "");
      const id = callId ? toolExecutionIds.get(callId) : undefined;
      if (!id) return;
      const isError = event.type === StreamEventType.TOOL_ERROR;
      this.store.finishToolExecution(
        id,
        isError ? "error" : "succeeded",
        isError ? null : String(payload.result_summary ?? payload.result ?? ""),
        isError ? String(payload.error_text ?? payload.error ?? "") : null,
      );
      toolExecutionIds.delete(callId);
    }
  }

  private static serverConfigsFor(servers: { name: string; toLangchainConfig(): Record<string, unknown> }[]): Record<string, Record<string, unknown>> {
    const configs: Record<string, Record<string, unknown>> = {};
    for (const server of servers) configs[server.name] = server.toLangchainConfig();
    return configs;
  }

  private static builtinAndMiddlewareToolCount(rag: { knowledgeBases: { enabled: boolean; name: string }[] }): number {
    const builtin = 3 + (rag.knowledgeBases.some((kb) => kb.enabled && kb.name) ? 1 : 0);
    const middleware = 8;
    return builtin + middleware;
  }

  async close(): Promise<void> {
    this.cancelCurrentRun();
    await this.pool.disconnect();
  }
}

function providerToDto(provider: ProviderConfig): ModelProviderDTO {
  return new ModelProviderDTO({
    id: null,
    name: provider.name,
    modelName: provider.modelName,
    apiMode: provider.apiMode,
    apiKey: provider.apiKey,
    baseUrl: provider.baseUrl,
    temperature: provider.temperature,
    topP: provider.topP,
    maxContextTokens: provider.maxContextTokens,
    enabled: provider.enabled,
  });
}
